tpblite/models/utils.py

```diff
@@ -1,57 +1,80 @@
+from typing import Tuple, Type
 import random
 from urllib.request import Request, urlopen
 import urllib.error
 from purl import URL as pURL
 
 
-class QueryParser(object):
-    '''
-    Query object capable of getting html response given 
-    a search query and other parameters.
-    '''
-    def __init__(self, query, base_url, page, order, category):
+class QueryParser:
+    """Query object capable of getting html response given a search query and other
+    parameters.
+    """
+
+    # PirateBay URL to use for queries
+    base_url: str
+
+    # Compiled search string used to query the PirateBay URL
+    url: str
+
+    def __init__(self, base_url: str, segments: Tuple[str, ...]):
         self.base_url = base_url
-        segments = ('search', query, str(page), str(order), str(category))
         self.url = URL(base_url, segments)
         try:
             self.html_source = self._sendRequest()
         except urllib.error.URLError:
-            raise ConnectionError('Could not establish connection wtih {}'.format(self.base_url))
-     
+            raise ConnectionError(
+                "Could not establish connection wtih {}".format(self.base_url)
+            )
+
+    @classmethod
+    def from_search(
+        cls, query: str, base_url: str, page: int, order: int, category: int
+    ):
+        segments = ("search", query, str(page), str(order), str(category))
+        return cls(base_url, segments)
+
+    @classmethod
+    def from_browse(cls, base_url: str, category: int, page: int, order: int):
+        print("browsing")
+        segments = ("browse", str(category), str(page), str(order), "0")
+
+        return cls(base_url, segments)
+
     def _sendRequest(self):
         req = Request(self.url, headers=headers())
         return urlopen(req).read()
 
-def URL(base, segments):
+
+def URL(base: str, segments: Tuple[str, ...]) -> str:
```

Comment:
I'm not actually super familiar with type hints, but what does the `...` part of `Tuple[str, ...]` refer to/do?

Reply:
To answer my own question, from the typing docs:

> To specify a variable-length tuple of homogeneous type, use literal ellipsis, e.g. `Tuple[int, ...]`. A plain Tuple is equivalent to `Tuple[Any, ...]`, and in turn to tuple.

Reply:
Exactly ;) This is because the segments tuple takes different lengths for a browse or search query